### DISCUSSION TITLE
Shorten function with bool variable?

### DISCUSSION THREAD

**Discussion by doanhquang2005** [dotnet/csharplang#5080] [Q&A]
Ex: `bool IsProcess;
void function(){
if(IsProcess) return;
IsProcess = true;
// TODO
IsProcess = false;
}`
![image](https://user-images.githubusercontent.com/30622119/129855987-d492a4df-98ee-4f81-8ef4-6b328224f61a.png)
Do we have any shorthand for this case? I have many such commands in my project

**Comment by jnm2** [comment]
That's pretty dangerous. If the method throws an exception, you'll be stuck with `IsProcess` being true forever.

This is the usual pattern:

```cs
private bool isProcessing;

private void Function()
{
    if (isProcessing) return;
    isProcessing = true;
    try
    {
        // ...
    }
    finally
    {
        isProcessing = false;
    }
}
```

This can be encapsulated in various ways to avoid repetition of the full pattern.

**Reply by huoyaoyuan** [reply]
And to be thread safe, you need `lock`, `Semaphore` or `CompareExchange`. The library may contain some encapsulation for the patterns, but they are too complex for language.

**Reply by jnm2** [reply]
Very good point. Here's the thread-safe version of the same code, actually not requiring CompareExchange:

```cs
private int isProcessing;

private void Function()
{
    if (Interlocked.Exchange(ref isProcessing, 1) != 0) return;
    try
    {
        // ...
    }
    finally
    {
        Volatile.Write(ref isProcessing, 0);
    }
}
```
